applications/HDF5Application/custom_io/hdf5_file.cpp

```diff
@@ -971,19 +888,294 @@ void File::SetFileDriver(const std::string& rDriver, hid_t FaplId) const
 }
 
-template void File::WriteAttribute(const std::string& rObjectPath, const std::string& rName, int Value);
-template void File::WriteAttribute(const std::string& rObjectPath, const std::string& rName, double Value);
-template void File::WriteAttribute(const std::string& rObjectPath, const std::string& rName, const Vector<int>& rValue);
-template void File::WriteAttribute(const std::string& rObjectPath, const std::string& rName, const Vector<double>& rValue);
-template void File::WriteAttribute(const std::string& rObjectPath, const std::string& rName, const Matrix<int>& rValue);
-template void File::WriteAttribute(const std::string& rObjectPath, const std::string& rName, const Matrix<double>& rValue);
+template<class TDataType, File::DataTransferMode TDataTransferMode>
+void File::WriteDataSetImpl(
+    const std::string& rPath,
+    const TDataType& rData,
+    WriteInfo& rInfo)
+{
+    KRATOS_TRY
+
+    using type_trait = DataTypeTraits<TDataType>;
+
+    static_assert(type_trait::IsContiguous, "HDF5File can only write contiguous data sets.");
+
+    constexpr auto local_dimension = type_trait::Dimension;
+
+    constexpr auto global_dimension = (local_dimension == 1 ? 1 : 2);
+
+    BuiltinTimer timer;
+
+    // Create any missing subpaths.
+    auto pos = rPath.find_last_of('/');
+    if (pos != 0) {// Skip if last '/' is root.
+        std::string sub_path = rPath.substr(0, pos);
+        AddPath(sub_path);
+    }
+
+    // Initialize data space dimensions.
+    std::vector<hsize_t> local_shape(local_dimension);
+    type_trait::Shape(rData, local_shape.data(), local_shape.data() + local_dimension);
+
+    const hsize_t number_of_local_primitive_data_values = type_trait::Size(rData);
+
+    const auto& r_data_communicator = GetDataCommunicator();
+    // get the maximized dimensions of the underlying data. Max is taken because,
+    // there can be empty ranks which will give wrong sizes in the case of dynamic
+    // data types.
+    if constexpr(local_dimension >= 2) {
+        if constexpr(type_trait::template IsDimensionDynamic<1>()) {
+            // this is the matrix version. Hence it is better to get the max size
+            // from all ranks.
+            const auto max_size = r_data_communicator.MaxAll(static_cast<unsigned int>(local_shape[1]));
+
+            // now check every non-empty ranks have the same sizes since this dimension
+            // is a dynamic dimension.
+            KRATOS_ERROR_IF(number_of_local_primitive_data_values > 0 && max_size != local_shape[1])
+                << "Mismatching shapes found in different ranks. All ranks should have the same shapes in data sets.";
+
+            local_shape[1] = max_size;
+        }
+    }
+
+    // local_reduced_shape holds the max 2d flattened shape if the local_shape dimensions
+    // are higher than 2.
+    std::vector<hsize_t> global_shape(global_dimension, 0), local_reduced_shape(global_dimension, 0), local_shape_start(global_dimension, 0);
```

Comment:
`global_dimension` is a compile time constant, so these could be `std::array`s.

Reply:
I would keep some, otherwise I have to template some callee methods as well.

Reply:
ok done... removed the callee method totally. :D.